src/macaron/repo_finder/repo_verifier.py

```diff
@@ -0,0 +1,300 @@
+        "bitbucket",
+        "gitee",
+    }
+
+    def __init__(
+        self,
+        namespace: str,
+        name: str,
+        version: str,
+        claimed_repo_url: str,
+        claimed_repo_fs: str,
+        build_tool: BaseBuildTool,
+    ):
+        self.namespace = namespace
+        self.name = name
+        self.version = version
+        self.claimed_repo_url = claimed_repo_url
+        self.claimed_repo_fs = claimed_repo_fs
+        self.build_tool = build_tool
+
+    def verify_claimed_repo(self) -> RepositoryVerificationResult:
+        """Verify whether the claimed repository links back to the artifact."""
+        default_res = RepositoryVerificationResult(
+            status=RepositoryVerificationStatus.UNKNOWN, reason="unsupported_type", build_tool=self.build_tool
+        )
+
+        match self.build_tool:
+            case Maven():
+                git_ns_ver = self._verify_maven_git_ns()
+                if git_ns_ver.status == RepositoryVerificationStatus.PASSED:
+                    return git_ns_ver
+                return self._verify_maven()
+            case Gradle():
+                git_ns_ver = self._verify_maven_git_ns()
+                if git_ns_ver.status == RepositoryVerificationStatus.PASSED:
+                    return git_ns_ver
+                return self._verify_gradle()
+            # TODO: add verifiers for other types
+            case Poetry():
+                return default_res
+            case Pip():
+                return default_res
+            case Docker():
+                return default_res
+            case NPM():
+                return default_res
+            case Yarn():
+                return default_res
+            case Go():
+                return default_res
+            case _:
+                raise NotImplementedError(f"Unsupported build tool: {self.build_tool}")
+
+    def _same_group(self, g1: str, g2: str) -> bool:
+        if g1 == g2:
+            return True
+
+        g1_parts = g1.split(".")
+        g2_parts = g2.split(".")
+        if min(len(g1_parts), len(g2_parts)) < 2:
+            return False
+
+        if (g1_parts[0] in {"io", "com"} and g1_parts[1] in self._known_maven_namespaces) or (
+            g2_parts[0] in {"io", "com"} and g2_parts[1] in self._known_maven_namespaces
+        ):
+            if len(g1_parts) >= 3 and len(g2_parts) >= 3:
+                return all(g1_parts[i] == g2_parts[i] for i in range(3))
+            return False
+
+        for i in range(2):
+            if g1_parts[i] != g2_parts[i]:
+                return False
+
+        return True
+
+    @staticmethod
+    def _bfs_walk(root_dir: Path, filename: str) -> Path | None:
+        if not os.path.exists(root_dir) or not os.path.isdir(root_dir):
+            return None
+
+        queue: deque[Path] = deque()
+        queue.append(Path(root_dir))
+        while queue:
+            current_dir = queue.popleft()
+
+            # don't look through non-main directories
```

Comment:
Please start all comments with a capital letter (where applicable), and end them with a period.